module/zfs/bpobj.c

```diff
@@ -242,119 +282,200 @@ bpobj_iterate_impl(bpobj_t *bpo, bpobj_itor_t func, void *arg, dmu_tx_t *tx,
 			ASSERT3S(bpo->bpo_phys->bpo_num_blkptrs, >=, 0);
 		}
 	}
+	if (free) {
+		propagate_space_reduction(bpi, freed, comp_freed,
+		    uncomp_freed, tx);
+		VERIFY0(dmu_free_range(bpo->bpo_os,
+		    bpo->bpo_object,
+		    bpo->bpo_phys->bpo_num_blkptrs * sizeof (blkptr_t),
+		    DMU_OBJECT_END, tx));
+	}
 	if (dbuf) {
 		dmu_buf_rele(dbuf, FTAG);
 		dbuf = NULL;
 	}
-	if (free) {
-		VERIFY3U(0, ==, dmu_free_range(bpo->bpo_os, bpo->bpo_object,
-		    (i + 1) * sizeof (blkptr_t), DMU_OBJECT_END, tx));
-	}
-	if (err || !bpo->bpo_havesubobj || bpo->bpo_phys->bpo_subobjs == 0)
-		goto out;
-
-	ASSERT(bpo->bpo_havecomp);
-	err = dmu_object_info(bpo->bpo_os, bpo->bpo_phys->bpo_subobjs, &doi);
-	if (err) {
-		mutex_exit(&bpo->bpo_lock);
-		return (err);
-	}
-	ASSERT3U(doi.doi_type, ==, DMU_OT_BPOBJ_SUBOBJ);
-	epb = doi.doi_data_block_size / sizeof (uint64_t);
-
-	for (i = bpo->bpo_phys->bpo_num_subobjs - 1; i >= 0; i--) {
-		uint64_t *objarray;
-		uint64_t offset, blkoff;
-		bpobj_t sublist;
-		uint64_t used_before, comp_before, uncomp_before;
-		uint64_t used_after, comp_after, uncomp_after;
+	return (err);
+}
 
-		offset = i * sizeof (uint64_t);
-		blkoff = P2PHASE(i, epb);
+/*
+ * Given an initial bpo, start by freeing the BPs that are directly referenced
+ * by that bpo. If the bpo has subobjs, read in its last subobj and push the
+ * subobj to our stack. By popping items off our stack, eventually we will
+ * encounter a bpo that has no subobjs.  We can free its bpobj_info_t, and if
+ * requested also free the now-empty bpo from disk and decrement
+ * its parent's subobj count. We continue popping each subobj from our stack,
+ * visiting its last subobj until they too have no more subobjs, and so on.
+ */
+static int
+bpobj_iterate_impl(bpobj_t *initial_bpo, bpobj_itor_t func, void *arg,
+    dmu_tx_t *tx, boolean_t free)
+{
+	list_t stack;
+	bpobj_info_t *bpi;
+	int err = 0;
 
-		if (dbuf == NULL || dbuf->db_offset > offset) {
-			if (dbuf)
-				dmu_buf_rele(dbuf, FTAG);
-			err = dmu_buf_hold(bpo->bpo_os,
-			    bpo->bpo_phys->bpo_subobjs, offset, FTAG, &dbuf, 0);
-			if (err)
+	/*
+	 * Create a "stack" for us to work with without worrying about
+	 * stack overflows. Initialize it with the initial_bpo.
+	 */
+	list_create(&stack, sizeof (bpobj_info_t),
+	    offsetof(bpobj_info_t, bpi_node));
+	mutex_enter(&initial_bpo->bpo_lock);
+	list_insert_head(&stack, bpi_alloc(initial_bpo, NULL, 0, 0));
+
+	while ((bpi = list_head(&stack)) != NULL) {
+		bpobj_t *bpo = bpi->bpi_bpo;
+
+		ASSERT0(err);
+		ASSERT3P(bpo, !=, NULL);
+		ASSERT(MUTEX_HELD(&bpo->bpo_lock));
+		ASSERT(bpobj_is_open(bpo));
+
+		if (free)
+			dmu_buf_will_dirty(bpo->bpo_dbuf, tx);
+
+		if (bpi->bpi_visited == B_FALSE) {
+			err = bpobj_iterate_blkptrs(bpi, func, arg, tx, free);
+			bpi->bpi_visited = B_TRUE;
+			if (err != 0)
 				break;
 		}
+		/*
+		 * We've finished with this bpo's directly-referenced BP's.
+		 * If it has no subobjs, we are done processing it. We can free
```

Comment:
How about `If we have also processed all its subobjs, we are done processing it.`